src/http.handler.ts

```diff
@@ -5,77 +5,81 @@ import {
 } from "./http.response.handler";
 import { httpErrorHandler } from "./http.error.handler";
 import { HttpStatusCode } from "./enum";
-import { APIGatewayEvent, Context, APIGatewayProxyResult } from "aws-lambda";
-import { HttpErrorException } from "./exceptions";
-import { createLoggingHandler, ErrorHandlingType } from "./logging.handler";
-
-export type APIGatewayJsonEvent<
-  T extends { [s: string]: any }
-> = APIGatewayEvent & {
-  json: T;
-};
-
-export type HttpHandlerFunction<T, R> = (
-  event?: APIGatewayEvent | APIGatewayJsonEvent<T>,
-  context?: Context,
-) => Promise<void | Partial<APIGatewayProxyResult> | APIGatewayProxyResult | R>;
+import {
+  APIGatewayEvent,
+  Context,
+  APIGatewayProxyResult,
+  APIGatewayProxyHandler,
+  Callback,
+} from "aws-lambda";
+import { createLoggingHandler } from "./logging.handler";
+import { JSONParse } from "./utils/json.parse";
+import { HttpHandlerFunctionOptions } from "./interfaces";
 
 /**
  * A universal wrapper function response hander for aws handlers
- *
- * @param fn your custom handler function
- * @param defaultStatus Http Status Code
- * @param errorHandlingOptions
  */
-export const httpHandler = <T extends { [s: string]: any }, R extends any>(
-  fn: HttpHandlerFunction<T, R>,
-  defaultStatus: HttpStatusCode = HttpStatusCode.OK,
-  errorHandlingOptions: ErrorHandlingType = [HttpStatusCode.BAD_REQUEST, HttpStatusCode.NETWORK_READ_TIMEOUT],
-): ((
-  event: APIGatewayEvent,
-  context: Context,
-) => Promise<APIGatewayProxyResult>) => (
+export const httpHandler = <RequestType extends any, ResponseType extends any>(
+  handlerOptions: HttpHandlerFunctionOptions<RequestType, ResponseType>,
+): APIGatewayProxyHandler => (
   event: APIGatewayEvent,
   context: Context,
+  callback: Callback<APIGatewayProxyResult>,
 ): Promise<APIGatewayProxyResult> => {
-  const loggingHandler = createLoggingHandler();
+  const options: HttpHandlerFunctionOptions<RequestType, ResponseType> =
```

Comment:
To avoid having to all the checks e.g.

```options.errorHandler ? blah : errorHandler```

We could instead write:

```suggestion
  const options: HttpHandlerFunctionOptions<RequestType, ResponseType> = {
    logger: createLoggingHandler(),
    serialise: { 
      input: options?.serialise?.input(event) || JSONParse(event),
      output: 
    },
    validator: (input: any) => input as RequestType,
    ...handlerOptions,
  }
```